header/sync/sync.go

```diff
@@ -236,44 +231,104 @@ func (s *Syncer) doSync(ctx context.Context, fromHead, toHead *header.ExtendedHe
 
 // processHeaders gets and stores headers starting at the given 'from' height up to 'to' height -
 // [from:to]
-func (s *Syncer) processHeaders(ctx context.Context, from, to uint64) (int, error) {
-	headers, err := s.findHeaders(ctx, from, to)
+func (s *Syncer) processHeaders(ctx context.Context, fromHeader *header.ExtendedHeader, to uint64) error {
+	headers, err := s.requestHeaders(ctx, fromHeader, to)
 	if err != nil {
-		return 0, err
+		return err
+	}
+	_, err = s.store.Append(ctx, headers...)
+	return err
+}
+
+// requestHeaders checks headers in pending cache that apply request range.
+// If some headers are missing, it starts requesting them from the network.
+// All headers that will be received are verified to be contiguous.
+func (s *Syncer) requestHeaders(
+	ctx context.Context,
+	fromHeader *header.ExtendedHeader,
+	to uint64,
+) ([]*header.ExtendedHeader, error) {
+	amount := to - uint64(fromHeader.Height)
+	cached, ok := s.checkCache(fromHeader, to)
+	if !ok {
+		// request full range if cache is empty
+		return s.requestRange(ctx, fromHeader, to)
+	}
+
+	out := make([]*header.ExtendedHeader, 0, amount)
+	for _, headers := range cached {
+		if fromHeader.Height+1 == headers[0].Height {
+			// apply cache
+			out = append(out, headers...)
+			// set new header to count from
+			fromHeader = out[len(out)-1]
+			continue
+		}
+		// make an external request
+		h, err := s.requestRange(ctx, fromHeader, uint64(headers[0].Height)-1)
+		if err != nil {
+			return nil, err
+		}
+		// apply received headers + headers from the range
+		out = append(out, append(h, headers...)...)
+		fromHeader = out[len(out)-1]
 	}
 
-	return s.store.Append(ctx, headers...)
+	// ensure that we have all requested headers
+	if uint64(len(out)) == amount {
+		return out, nil
+	}
+	// make on more external request in case if `to` is bigger than the
+	// last cached header
+	h, err := s.requestRange(ctx, fromHeader, to)
+	if err != nil {
+		return nil, err
+	}
+	out = append(out, h...)
+	return out, nil
 }
 
-// findHeaders gets headers from either remote peers or from local cache of headers received by
-// PubSub - [from:to]
-func (s *Syncer) findHeaders(ctx context.Context, from, to uint64) ([]*header.ExtendedHeader, error) {
-	amount := to - from + 1 // + 1 to include 'to' height as well
-	if amount > s.Params.MaxRequestSize {
-		to, amount = from+s.Params.MaxRequestSize, s.Params.MaxRequestSize
+// checkCache returns all headers sub-ranges of headers that could be found in between the requested range.
+func (s *Syncer) checkCache(fromHeader *header.ExtendedHeader, to uint64) ([][]*header.ExtendedHeader, bool) {
+	r := s.pending.FindAllRangesWithin(uint64(fromHeader.Height+1), to) // start looking for headers for the next height
+	if len(r) == 0 {
+		return nil, false
+	}
+
+	out := make([][]*header.ExtendedHeader, len(r))
+	for index, headersRange := range r {
```

Comment:
let's just call this more idiomatic `i` in general unless it's meant to represent something more meaningful.

```suggestion
	for i, headersRange := range r {
```